src/test/java/org/sonatype/nexus/repository/composer/internal/ComposerFormatAttributesExtractorTest.java

```diff
@@ -0,0 +1,269 @@
+    extends TestSupport
+{
+  private static final String EXPECTED_FIELDS[] = new String[]{
+      P_NAME, P_VERSION, P_DESCRIPTION, P_TYPE, P_KEYWORDS, P_HOMEPAGE, P_TIME, P_LICENSE, P_AUTHORS, P_SUPPORT_EMAIL,
+      P_SUPPORT_ISSUES, P_SUPPORT_FORUM, P_SUPPORT_WIKI, P_SUPPORT_SOURCE, P_SUPPORT_DOCS, P_SUPPORT_RSS
+  };
+
+  @Mock
+  private Blob blob;
+
+  @Mock
+  private TempBlob tempBlob;
+
+  @Test
+  public void extractInfoFromZipballWithJson() throws Exception {
+    try (InputStream in = getClass().getResourceAsStream("extractInfoFromZipballWithJson.zip")) {
+      when(tempBlob.getBlob()).thenReturn(blob);
+      when(blob.getInputStream()).thenReturn(in);
+
+      NestedAttributesMap attributesMap = new NestedAttributesMap("composer", new LinkedHashMap<>());
+      ComposerFormatAttributesExtractor underTest = new ComposerFormatAttributesExtractor();
+      underTest.extractFromZip(tempBlob, attributesMap);
+
+      assertThat(attributesMap.keys(), containsInAnyOrder(EXPECTED_FIELDS));
+      assertThat(attributesMap.get(P_NAME), is("vendor/project"));
+      assertThat(attributesMap.get(P_VERSION), is("1.2.3"));
+      assertThat(attributesMap.get(P_DESCRIPTION), is("Test description"));
+      assertThat(attributesMap.get(P_TYPE), is("library"));
+      assertThat((List<String>) attributesMap.get(P_KEYWORDS), containsInAnyOrder("keyword1", "keyword2", "keyword3"));
+      assertThat(attributesMap.get(P_HOMEPAGE), is("http://www.example.com/"));
+      assertThat(attributesMap.get(P_TIME), is("2008-05-15"));
+      assertThat(attributesMap.get(P_LICENSE), is("MIT"));
+      assertThat((List<String>) attributesMap.get(P_AUTHORS),
+          containsInAnyOrder(
+              "Author1 <author1@example.com> (http://www.example.com/author1)",
+              "Author2 <author2@example.com> (http://www.example.com/author2)"));
+      assertThat(attributesMap.get(P_SUPPORT_EMAIL), is("email@example.com"));
+      assertThat(attributesMap.get(P_SUPPORT_ISSUES), is("irc://irc.example.com/irc"));
+      assertThat(attributesMap.get(P_SUPPORT_FORUM), is("http://www.example.com/forum"));
+      assertThat(attributesMap.get(P_SUPPORT_WIKI), is("http://www.example.com/wiki"));
+      assertThat(attributesMap.get(P_SUPPORT_SOURCE), is("http://www.example.com/source"));
+      assertThat(attributesMap.get(P_SUPPORT_DOCS), is("http://www.example.com/docs"));
+      assertThat(attributesMap.get(P_SUPPORT_RSS), is("http://www.example.com/rss"));
+    }
+  }
+
+  @Test
+  public void extractInfoFromZipballWithoutJson() throws Exception {
+    try (InputStream in = getClass().getResourceAsStream("extractInfoFromZipballWithoutJson.zip")) {
+      when(tempBlob.getBlob()).thenReturn(blob);
+      when(blob.getInputStream()).thenReturn(in);
+
+      NestedAttributesMap attributesMap = new NestedAttributesMap("composer", new LinkedHashMap<>());
+      ComposerFormatAttributesExtractor underTest = new ComposerFormatAttributesExtractor();
```

Comment:
This is used in most tests so could go in setup.

Reply:
> This is used in most tests so could go in setup.

Overall I've been leaning more toward just keeping the declarations as close as possible to where they're used, but in this cases they do have a lot of commonality. Done in https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/11/commits/97b0aa140b325bb5be24906969c0f9a5a937c843 and https://github.com/sonatype-nexus-community/nexus-repository-composer/pull/11/commits/0e02ec7317a08e6e1c67c469b4990e14c3124684.

Reply:
This is more my own musings/findings rather than direction for anyone else but I've found that defining the "happy path" in the setup method but then keeping the setup for the deviant cases within the tests reduces duplication and makes the tests easier to read. It also makes the edge cases easier to spot because they're the ones with the "complicated" tests.